### DISCUSSION TITLE
DFPSR

### DISCUSSION THREAD

**Discussion by dcrawl** [Dawoodoz/DFPSR#114] [General]
Not long ago, I tried to create my own software render. It was a lot of work and in the end, the performance wasn't all that great. Of course, that's not why I built it to begin with. I wanted to get a better understanding of the graphics pipeline. But I also wanted to create a retro style dungeon crawler game based on it too. However, I just couldn't get it to the speed that I was hoping for. So I'm looking at your project, which as far as I can tell, is probably the most robust and complete software renderer I have found on Github -- possibly anywhere on the net.  So I can appreciate the amount of hard work you must have put into this.  I'm glad to see that it is still being maintained.

I  just hope I can figure out how to integrate it into my existing work.  Thanks for this project.

James

**Comment by Dawoodoz** [comment]
I could have a look at the code and give some integration advice if you have a link to the project.

Regards, David Piuva

**Comment by dcrawl** [comment]
Thanks for the offer.  At the moment, I'm having trouble with just compiling the Stable branch.  I'm just running the build_linux.sh (even though I'm on a Mac)  and I receive an error:

In file included from ./DFPSR/base/SafePointer.cpp:25:
./DFPSR/base/../api/stringAPI.h:149:9: error: call to 'string_toStreamIndented' is ambiguous
  149 |         return string_toStreamIndented(target, source, U"");
  
I'm guessing this has something to do with the template but I'm not a C++ expert.  So I'm not 100% sure how to fix it.  I'm on an M1 Mac, if that helps.  I get the same error whether I use clang++ or g++.

James

**Reply by Dawoodoz** [reply]
It probably means that the type given to the string_toStream template does not know which overload of string_toStreamIndented to call next due to multiple matching types for T. If you know which type was given as T, it might tell which types became ambiguous on Macintosh. You could try commenting out some of the overloads above and see if one is redundant, such as the string_toStreamIndented function that takes char* and see if it falls back on implicit conversion to std::string. There is also char32_t* which is implicitly converted to ReadableString.

**Reply by Dawoodoz** [reply]
You can try to add the line
CompilerFlag "-stdlib=libstdc++"
in the project's *.DsrProj file, so that GCC uses the same standard library for resolving overloads as on Linux.

The same flag will also be needed in tools/builder/buildProject.sh where the build system itself is compiled.

**Reply by Dawoodoz** [reply]
Then you will need to install an X11 compatibility layer if using a MacOS version without native X11 support.

Pulseaudio has no backend yet, so comment out any Sound flag to build muted programs.

**Reply by Dawoodoz** [reply]
According to the C++ standard specification, a direct match should be preferred over implicit conversion, so this might be a trait bug in the libc++ standard library implementation.

**Reply by Dawoodoz** [reply]
There was a bug in GCC that missed undeclared use of std::string because it is somehow hardcoded in the compiler, so moving #include < string >; from stringAPI.cpp to stringAPI.h might help.

**Comment by dcrawl** [comment]
I moved everything from stringAPI.cpp except the stringAPI.h and fileAPI.h to the stringAPI.h.  But I still get the same error.  I also changed compilers to clang++ and still receive the same error.  The complete list of output from the compile script is:

sh ./build_linux.sh
Running build_linux.sh 
Change CPP_COMPILER_PATH in /buildProject.sh if you are not using /usr/bin/clang++ as your compiler.
Running buildProject.sh Wizard.DsrProj Linux
BUILDER_FOLDER = /Users/redspark/Documents/GitHub/DFPSR/Source/tools/builder
BUILDER_EXECUTABLE = /Users/redspark/Documents/GitHub/DFPSR/Source/tools/builder/builder
Building the Builder build system for first time use.
In file included from /Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/SafePointer.cpp:25:
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:153:9: error: call to 'string_toStreamIndented' is ambiguous
  153 |         return string_toStreamIndented(target, source, U"");
      |                ^~~~~~~~~~~~~~~~~~~~~~~
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:350:2: note: in instantiation of function template specialization 'dsr::string_toStream<unsigned long>' requested here
  350 |         string_toStream(target, value);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:355:2: note: in instantiation of function template specialization 'dsr::string_append<unsigned long>' requested here
  355 |         string_append(target, head);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:356:2: note: in instantiation of function template specialization 'dsr::string_append<unsigned long, const char32_t (&)[5], unsigned long &, const char32_t (&)[2]>' requested here
  356 |         string_append(target, tail...);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/SafePointer.cpp:46:3: note: in instantiation of function template specialization 'dsr::string_append<const char32_t *, unsigned long, const char32_t (&)[5], unsigned long, const char32_t (&)[2]>' requested here
   46 |                 string_append(message, U"|  Region: ", (uintptr_t)regionStart, U" to ", (uintptr_t)regionEnd, U"\n");
      |                 ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:138:9: note: candidate function
  138 | String& string_toStreamIndented(String& target, const float& value, const ReadableString& indentation);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:139:9: note: candidate function
  139 | String& string_toStreamIndented(String& target, const double& value, const ReadableString& indentation);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:140:9: note: candidate function
  140 | String& string_toStreamIndented(String& target, const int64_t& value, const ReadableString& indentation);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:141:9: note: candidate function
  141 | String& string_toStreamIndented(String& target, const uint64_t& value, const ReadableString& indentation);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:142:9: note: candidate function
  142 | String& string_toStreamIndented(String& target, const int32_t& value, const ReadableString& indentation);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:143:9: note: candidate function
  143 | String& string_toStreamIndented(String& target, const uint32_t& value, const ReadableString& indentation);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:144:9: note: candidate function
  144 | String& string_toStreamIndented(String& target, const int16_t& value, const ReadableString& indentation);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:145:9: note: candidate function
  145 | String& string_toStreamIndented(String& target, const uint16_t& value, const ReadableString& indentation);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:146:9: note: candidate function
  146 | String& string_toStreamIndented(String& target, const int8_t& value, const ReadableString& indentation);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:147:9: note: candidate function
  147 | String& string_toStreamIndented(String& target, const uint8_t& value, const ReadableString& indentation);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:153:9: error: call to 'string_toStreamIndented' is ambiguous
  153 |         return string_toStreamIndented(target, source, U"");
      |                ^~~~~~~~~~~~~~~~~~~~~~~
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:350:2: note: in instantiation of function template specialization 'dsr::string_toStream<long>' requested here
  350 |         string_toStream(target, value);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:355:2: note: in instantiation of function template specialization 'dsr::string_append<long>' requested here
  355 |         string_append(target, head);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:356:2: note: in instantiation of function template specialization 'dsr::string_append<long, const char32_t (&)[8]>' requested here
  356 |         string_append(target, tail...);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/SafePointer.cpp:47:3: note: in instantiation of function template specialization 'dsr::string_append<const char32_t *, long, const char32_t (&)[8]>' requested here
   47 |                 string_append(message, U"|  Region size: ", (intptr_t)(regionEnd - regionStart), U" bytes\n");
      |                 ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:138:9: note: candidate function
  138 | String& string_toStreamIndented(String& target, const float& value, const ReadableString& indentation);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:139:9: note: candidate function
  139 | String& string_toStreamIndented(String& target, const double& value, const ReadableString& indentation);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:140:9: note: candidate function
  140 | String& string_toStreamIndented(String& target, const int64_t& value, const ReadableString& indentation);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:141:9: note: candidate function
  141 | String& string_toStreamIndented(String& target, const uint64_t& value, const ReadableString& indentation);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:142:9: note: candidate function
  142 | String& string_toStreamIndented(String& target, const int32_t& value, const ReadableString& indentation);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:143:9: note: candidate function
  143 | String& string_toStreamIndented(String& target, const uint32_t& value, const ReadableString& indentation);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:144:9: note: candidate function
  144 | String& string_toStreamIndented(String& target, const int16_t& value, const ReadableString& indentation);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:145:9: note: candidate function
  145 | String& string_toStreamIndented(String& target, const uint16_t& value, const ReadableString& indentation);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:146:9: note: candidate function
  146 | String& string_toStreamIndented(String& target, const int8_t& value, const ReadableString& indentation);
      |         ^
/Users/redspark/Documents/GitHub/DFPSR/Source/DFPSR/base/../api/stringAPI.h:147:9: note: candidate function
  147 | String& string_toStreamIndented(String& target, const uint8_t& value, const ReadableString& indentation);
      |         ^
2 errors generated.
Failed building the Builder build system, which is needed to build your project!

**Reply by Dawoodoz** [reply]
Installing libc++ now, so that I can try compiling with the same standard library implementation.

**Reply by Dawoodoz** [reply]
The stable branch would not link with libc++ on Linux, because std::shared_ptr was incompatible somehow. You could try pulling and compiling the master branch, which has cleaned up some things in the string API.

**Reply by Dawoodoz** [reply]
If the master branch gives the same error, try removing all printing of unsigned integers. uint8_t...uint64_t

**Reply by Dawoodoz** [reply]
Now I could compile using libc++ on Linux, but by selecting the standard library for both compiling and linking, but it only works with more recent branches.

**Reply by Dawoodoz** [reply]
Maybe I can download the MacOS SDK and just cross compile without executing.

**Reply by dcrawl** [reply]
I think you can cross compile C++ with Zig, if that helps.

**Reply by Dawoodoz** [reply]
Tried both OSXCross with MacOS SDK and Zig. The usr/include/c++ folder was missing in MacOS SDK 10.14 and 10.15, so I did not get what I needed to cross compile with OSXCross. Zig could not find the standard C++ headers.

Maybe I have to find a decent pre-owned Macintosh and try to compile locally.

**Reply by Dawoodoz** [reply]
Now I got compilation errors from a MacOS server by setting up CI on Github.

**Reply by Dawoodoz** [reply]
Now it compiles on MacOS but it looks as if the mutex is freed before the memory pool is done using the mutex, so it crashes after passing a test.

**Reply by Dawoodoz** [reply]
All tests passed on MacOS with the master branch. You can now try to run an application with an X11 compatibility tool, compiled without sound.

**Reply by dcrawl** [reply]
Thanks.  I'll give it a try.  I'm back to work now and don't have as much time as I did to code, unfortunately.  Thanks for digging into that.

**Reply by Dawoodoz** [reply]
Got X11 running with XQuartz and M4 Pro gave 3D performance equivalent to a dedicated graphics card.

**Comment by Dawoodoz** [comment]
It is now possible to develop fullscreen games with sound on the latest version of MacOS.
